### ISSUE TITLE
question about what guarnatees the correctness of the c11_atomics atomic_flag test

### ISSUE THREAD

**Issue by karolherbst** [KhronosGroup/OpenCL-CTS#2052]
This is the kernel being executed in the first `atomic_flag` test, but I'm actually wondering what in the OpenCL spec guarnatees that this executes as expected by the CTS.

More specifically, what prevents a thread to race with a thread from a different subgroup on `destMemory[cnt]`. One thread could enter the iteration for e.g. `cnt == 5`, while another thread just reached `atomic_flag_clear_explicit(&destMemory[cnt],` for the same `cnt`, which means that two threads will enter the critical section, just the second one won't do anything.

I'm seeing this behavior with rusticl on zink on radv and at the moment it's not clear to me if that's my or the CTS' bug. Removing the `atomic_flag_clear_explicit(&destMemory[cnt], ` makes only one thread execute the "criticial section" for each value of `cnt`, but that fails the test later.

So what's actually guaranteeing the correctness of this test here? Or would this test need to be rewritten? I'm also mildly wondering what clvk did in order to pass this test or if that was never a problem in the first place. I can probably come up with a fix to guarantee that behavior, I'm just wondering if there is some undefined behavior at play here.

```c
Program source:
__kernel void test_atomic_kernel(uint threadCount, uint numDestItems, __global int *finalDest, __global int *oldValues, volatile __local atomic_flag *destMemory)
{
  uint  tid = get_global_id(0);


              // initialize atomics not reachable from host (first thread
              // is doing this, other threads are waiting on barrier)
              if(get_local_id(0) == 0)
                for(uint dstItemIdx = 0; dstItemIdx < numDestItems; dstItemIdx++)
                {
                  if(finalDest[dstItemIdx])
                    atomic_flag_test_and_set_explicit(destMemory+dstItemIdx,
                                                      memory_order_relaxed,
                                                      memory_scope_work_group);
                  else
                    atomic_flag_clear_explicit(destMemory+dstItemIdx,
                                               memory_order_relaxed,
                                               memory_scope_work_group);    }
  barrier(CLK_LOCAL_MEM_FENCE);

  uint cnt, stop = 0;
  for(cnt = 0; !stop && cnt < threadCount; cnt++) // each thread must find critical section where it is the first visitor
  {
    bool set = atomic_flag_test_and_set_explicit(&destMemory[cnt], memory_order_relaxed, memory_scope_work_group);
    atomic_work_item_fence(CLK_LOCAL_MEM_FENCE | CLK_GLOBAL_MEM_FENCE, memory_order_acquire,memory_scope_work_group);
    if (!set)
    {
      uint csIndex = get_enqueued_local_size(0)*get_group_id(0)+cnt;
      // verify that thread is the first visitor
      if(oldValues[csIndex] == 1000000000)
      {
        oldValues[csIndex] = tid; // set the winner id for this critical section
        stop = 1;
      }
      atomic_work_item_fence(CLK_LOCAL_MEM_FENCE | CLK_GLOBAL_MEM_FENCE, memory_order_release,memory_scope_work_group);
      atomic_flag_clear_explicit(&destMemory[cnt], memory_order_relaxed, memory_scope_work_group);
    }
  }

  // Copy final values to host reachable buffer
  barrier(CLK_LOCAL_MEM_FENCE);
  if(get_local_id(0) == 0) // first thread in workgroup
    for(uint dstItemIdx = 0; dstItemIdx < numDestItems; dstItemIdx++)

                finalDest[dstItemIdx] =
                    atomic_flag_test_and_set_explicit(destMemory+dstItemIdx,
                                                      memory_order_relaxed,
                                                      memory_scope_work_group);}
```

**Comment by karolherbst** [comment]
mhh, no, that would mean that something with `atomic_flag_test_and_set_explicit` is going wrong, but that's kinda of weird.. maybe there is a coherency issue somewhere, anyway, it's probably not a CTS bug then and I initially thought something else is going on.

**Comment by rjodinchr** [comment]
To follow up on the question regarding clvk: we are hitting failures on `c11_atomics/atomic_flag` under `memory_scope_device` on some devices.

### Issue
In `CBasicTestFlag`, `memory_scope_device` forces `groupSize = 16` across `threadCount = 1024` (from `gMaxDeviceThreads`) => 64 workgroups.

Because all 1024 threads start at `cnt = 0` in a single linear pass (`for(cnt = 0; !stop && cnt < threadCount; cnt++)`), threads heavily collide on index 0, 1, etc. Due to GPU warp scheduling disparities across compute units, threads that repeatedly lose the race advance forward and hit the loop bound (`cnt == 1024`) without ever claiming a section. This exits early and fails validation:
```text
ERROR: Critical section 494 not visited
ERROR: Reference values did not validate!
```

### Proposed Fix
Distribute the start offset by workgroup:
`startStr = (get_group_id(0) * get_enqueued_local_size(0)) % threadCount;`

- All 16 threads in workgroup 0 still start at 0 and heavily contend for sections 0..15 (preserving mutual exclusion & concurrency testing).
- Eliminates the artificial cross-workgroup stampede on index 0.
- Keeps the loop linear (`cnt++`) and bounded (`cnt < countStr`), with no risk of infinite loops.

```diff
diff --git a/test_conformance/c11_atomics/test_atomics.cpp b/test_conformance/c11_atomics/test_atomics.cpp
index 72d7334..1b40286 100644
--- a/test_conformance/c11_atomics/test_atomics.cpp
+++ b/test_conformance/c11_atomics/test_atomics.cpp
@@ -4592,9 +4592,15 @@ public:
                 break;
         }
 
+        std::string countStr =
+            LocalMemory() ? "get_enqueued_local_size(0)" : "threadCount";
+        std::string startStr = LocalMemory()
+            ? "0"
+            : "(get_group_id(0) * get_enqueued_local_size(0)) % threadCount";
+
         std::string program =
             "  uint cnt, stop = 0;\n"
-            "  for(cnt = 0; !stop && cnt < threadCount; cnt++) // each thread "
+            "  for(cnt = " + startStr + "; !stop && cnt < " + countStr + "; cnt++) // each thread "
             "must find critical section where it is the first visitor\n"
             "  {\n"
             "    bool set = atomic_flag_test_and_set"
```

With this change, the test passes reliably on all the platforms I've tested.
What do folks think about this approach? Would a PR along these lines be welcome?

**Comment by karolherbst** [comment]
> Distribute the start offset by workgroup: `startStr = (get_group_id(0) * get_enqueued_local_size(0)) % threadCount;`
> 

I mean the entire purpose of the test is to verify locking/atomic semantics, so getting rid of the 
concurrent accesses will surely make the test pass, but then also make it entirely meaningless.

If that causes cross subgroups to race on the same value, we should figure out why that is and not hide the problem?

**Comment by rjodinchr** [comment]
I'm not proposing to get rid of concurrent access, only cross-workgroup collisions, to reduce the extreme contention. Concurrency is fully preserved within each workgroup: all 16 threads in a group still start at the exact same index on iteration 0 and heavily race for the lock.

What happens on my platform is that with 1024 threads all racing for index 0 simultaneously, threads that repeatedly lose the race exhaust their loop iterations before ever managing to claim a section.

I tried forcing many more iterations, but that requires a huge number of retries, and under that level of contention the workload sometimes crashes (likely hitting a GPU watchdog or something). 

My patch just reduces that cross-workgroup pressure while keeping concurrent lock contention intact to maintain the purpose of the test.

**Comment by karolherbst** [comment]
But isn't the purpose of this test to verify that exactly this pattern doesn't cause issues? Like each invocation should only enter with a different index and I don't see as of now, why the test is wrong. If there are 1024 invocations, and 1024 loop iterations, then all invocations have to enter for a correct implementation.

And that's my current understanding. I was _wondering_ if the test relies on UB actually, because the issue I was seeing is that multiple invocations enter with the same index and I was wondering if the test needs stronger synchronization between the invocations instead.

But your explanation just describes a performance problem, which isn't relevant to the CTS? Like sure, the invocations access all the same data, but that's the entire point of the test that this pattern should actually be fine and that implementations should not misbehave and actually not cause infinite loops there.

**Comment by rjodinchr** [comment]
You are right. It is indeed a cache coherency issue on the clvk/clspv side.

Under the OpenCL memory model, a release operation on an atomic object synchronizes-with an acquire operation on the same object, establishing a happens-before relationship that guarantees prior non-atomic writes are visible. The test implementation relies directly on this specification guarantee.

Because Vulkan atomic operations alone do not flush or invalidate non-coherent per-core caches for ordinary memory, threads across different compute units were reading stale data. Multiple threads were mistakenly claiming the same critical section and exiting early, leaving later sections unvisited.

Emitting proper acquire and release memory barriers around the atomic flag operations in the compiler resolved the issue (https://github.com/google/clspv/pull/1652). The test now passes completely without any changes to the CTS.
